File: cortex/webgraph/service.py
from __future__ import annotations

from collections import deque
from pathlib import Path

from cortex.runtime_context import slugify
from cortex.enterprise.config import discover_enterprise_config_path, load_enterprise_config
from cortex.webgraph.cache import WebGraphCache
from cortex.webgraph.config import WebGraphConfig
from cortex.webgraph.contracts import WebGraphEdge, WebGraphMode, WebGraphNode, WebGraphNodeDetail, WebGraphSnapshot
from cortex.webgraph.episodic_source import EpisodicSource
from cortex.webgraph.graph_builder import GraphBuilder
from cortex.webgraph.semantic_source import SemanticSource
from cortex.workspace.layout import WorkspaceLayout


class WebGraphService:
    """High-level orchestration for building and querying Cortex webgraph snapshots."""

    def __init__(
        self,
        project_root: Path | None = None,
        *,
        config: WebGraphConfig | None = None,
        vault_path: Path | None = None,
        persist_dir: Path | None = None,
        semantic_source: SemanticSource | None = None,
        episodic_source: EpisodicSource | None = None,
        workspace_layout: WorkspaceLayout | None = None,
    ) -> None:
        self.project_root = (project_root or Path.cwd()).resolve()
        self._layout = workspace_layout or WorkspaceLayout.discover(self.project_root)
        self.config = config or WebGraphConfig.load(self.project_root, workspace_layout=self._layout)
        self.semantic_source = semantic_source or SemanticSource(self.project_root, vault_path=vault_path)
        self.episodic_source = episodic_source or EpisodicSource(self.project_root, persist_dir=persist_dir)
        self.cache = WebGraphCache(self.project_root, workspace_layout=self._layout)
        self.graph_builder = GraphBuilder(self.config)

    def build_snapshot(
        self,
        mode: WebGraphMode = "hybrid",
        *,
        use_cache: bool = True,
        scope: str | None = None,
    ) -> WebGraphSnapshot:
        fingerprint = self.cache.compute_fingerprint(
            vault_path=self.semantic_source.vault_path,
            episodic_path=self.episodic_source.persist_dir,
            episodic_count=self.episodic_source.store.count(),
            episodic_cache_token=self.episodic_source.store.cache_token,
            config_payload=self.config.model_dump(),
        )
        if use_cache:
            cached = self.cache.load_snapshot(mode, fingerprint)
            if cached is not None:
                return cached

        include_embeddings = mode == "hybrid"
        semantic_records = self.semantic_source.load_records(include_embeddings=include_embeddings)
        episodic_records = self.episodic_source.load_records(include_embeddings=include_embeddings)
        snapshot = self.graph_builder.build_snapshot(
            fingerprint=fingerprint,
            mode=mode,
            semantic_records=semantic_records,
            episodic_records=episodic_records,
        )
        project_id = slugify(self.project_root.name, fallback="project")
        snapshot = snapshot.model_copy(
            update={
                "nodes": [
                    node.model_copy(
                        update={"metadata": {"project_id": project_id, "scope": "local", **dict(node.metadata)}}
                    )
                    for node in snapshot.nodes
                ]
            }
        )

        snapshot = _append_enterprise_nodes(snapshot, self.project_root, project_id=project_id, workspace_layout=self._layout)
        snapshot = _filter_snapshot_by_scope(snapshot, scope)
        self.cache.store_snapshot(mode, snapshot)
        return snapshot

    def export_snapshot(
        self,
        output_path: Path | None = None,
        *,
        mode: WebGraphMode = "hybrid",
        use_cache: bool = True,
    ) -> Path:
        snapshot = self.build_snapshot(mode=mode, use_cache=use_cache)
        path = output_path or self.cache.snapshot_path(mode)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(snapshot.model_dump_json(indent=2), encoding="utf-8")
        return path

    def get_node_detail(self, node_id: str, *, mode: WebGraphMode = "hybrid") -> WebGraphNodeDetail:
        snapshot = self.build_snapshot(mode=mode)
        nodes_by_id = {node.id: node for node in snapshot.nodes}
        node = nodes_by_id[node_id]
        relations = [edge for edge in snapshot.edges if edge.source == node_id or edge.target == node_id]
        neighbor_ids = {
            edge.target if edge.source == node_id else edge.source
            for edge in relations
        }
        neighbors = [nodes_by_id[neighbor_id] for neighbor_id in sorted(neighbor_ids) if neighbor_id in nodes_by_id]
        return WebGraphNodeDetail(node=node, relations=relations, neighbors=neighbors)

    def resolve_node_path(self, node_id: str, *, mode: WebGraphMode = "hybrid") -> Path | None:
        detail = self.get_node_detail(node_id, mode=mode)
        rel_path = detail.node.rel_path
        if not rel_path:
            return None
        return (self.semantic_source.vault_path / rel_path).resolve()

    def get_subgraph(
        self,
        node_id: str,
        *,
        depth: int = 1,
        mode: WebGraphMode = "hybrid",
        edge_types: set[str] | None = None,
    ) -> WebGraphSnapshot:
        snapshot = self.build_snapshot(mode=mode)
        adjacency: dict[str, list[tuple[str, str]]] = {}
        for edge in snapshot.edges:
            if edge_types and edge.edge_type not in edge_types:
                continue
            adjacency.setdefault(edge.source, []).append((edge.target, edge.edge_type))
            adjacency.setdefault(edge.target, []).append((edge.source, edge.edge_type))

        visited = {node_id}
        queue: deque[tuple[str, int]] = deque([(node_id, 0)])
        while queue:
            current, current_depth = queue.popleft()
            if current_depth >= depth:
                continue
            for neighbor, _ in adjacency.get(current, []):
                if neighbor in visited:
                    continue
                visited.add(neighbor)
                queue.append((neighbor, current_depth + 1))

        nodes = [node for node in snapshot.nodes if node.id in visited]
        edges = [
            edge
            for edge in snapshot.edges
            if edge.source in visited and edge.target in visited
            and (not edge_types or edge.edge_type in edge_types)
        ]
        return snapshot.model_copy(
            update={
                "nodes": nodes,
                "edges": edges,
                "stats": snapshot.stats.model_copy(
                    update={"node_count": len(nodes), "edge_count": len(edges)},
                ),
            }
        )


def _append_enterprise_nodes(snapshot: WebGraphSnapshot, project_root: Path, *, project_id: str, workspace_layout: WorkspaceLayout | None = None) -> WebGraphSnapshot:
    layout = workspace_layout or WorkspaceLayout.discover(project_root)
    org_path = discover_enterprise_config_path(project_root, workspace_layout=layout)
    if org_path is None:
        return snapshot
    try:
        cfg = load_enterprise_config(project_root, required=True, path=org_path, workspace_layout=layout)
    except Exception:
        return snapshot
    if cfg is None:
        return snapshot

    org_node_id = "enterprise:org"
    proj_node_id = "enterprise:project"
    vault_node_id = "enterprise:vault"

    enterprise_nodes: list[WebGraphNode] = [
        WebGraphNode(
            id=org_node_id,
            node_type="enterprise_org",
            source="semantic",
            label=cfg.organization.name,
            summary=f"profile={cfg.organization.profile}, slug={cfg.organization.slug}",
            metadata={"project_id": project_id, "scope": "enterprise"},
        ),
        WebGraphNode(
            id=proj_node_id,
            node_type="enterprise_project",
            source="semantic",
            label=project_root.name,
            summary=str(project_root),
            metadata={"project_id": project_id, "scope": "enterprise"},
        ),
    ]

    enterprise_edges: list[WebGraphEdge] = [
        WebGraphEdge(
            id="enterprise:org->project",
            source=org_node_id,
            target=proj_node_id,
            edge_type="enterprise_owns_project",
            weight=1.0,
        ),
    ]

    vault_path = cfg.resolve_enterprise_vault_path(project_root, workspace_root=layout.workspace_root)
    if vault_path is not None:
        enterprise_nodes.append(
            WebGraphNode(
                id=vault_node_id,
                node_type="enterprise_vault",
                source="semantic",
                label=vault_path.name,
                summary=str(vault_path),
                rel_path=None,
                metadata={"project_id": project_id, "scope": "enterprise"},
            )
        )
        enterprise_edges.append(
            WebGraphEdge(
                id="enterprise:org->vault",
                source=org_node_id,
                target=vault_node_id,
                edge_type="enterprise_vault",
                weight=1.0,
            )
        )

    # Avoid duplicates if called multiple times.
    existing_ids = {n.id for n in snapshot.nodes}
    nodes = snapshot.nodes + [n for n in enterprise_nodes if n.id not in existing_ids]

    existing_edge_ids = {e.id for e in snapshot.edges}
    edges = snapshot.edges + [e for e in enterprise_edges if e.id not in existing_edge_ids]

    return snapshot.model_copy(
        update={
            "nodes": nodes,
            "edges": edges,
            "stats": snapshot.stats.model_copy(update={"node_count": len(nodes), "edge_count": len(edges)}),
        }
    )


def _filter_snapshot_by_scope(snapshot: WebGraphSnapshot, scope: str | None) -> WebGraphSnapshot:
    if scope is None or scope == "all":
        return snapshot
    if scope not in {"local", "enterprise"}:
        return snapshot

    allowed_ids = {
        n.id
        for n in snapshot.nodes
        if str(n.metadata.get("scope", "local")).strip().lower() == scope
    }
    nodes = [n for n in snapshot.nodes if n.id in allowed_ids]
    edges = [e for e in snapshot.edges if e.source in allowed_ids and e.target in allowed_ids]
    return snapshot.model_copy(
        update={
            "nodes": nodes,
            "edges": edges,
            "stats": snapshot.stats.model_copy(update={"node_count": len(nodes), "edge_count": len(edges)}),
        }
    )

File: rust/crates/cortex-webgraph-server/src/service.rs
//! Porteo de `cortex/webgraph/service.py` — orquestación de snapshots.
//!
//! GAP explícito (P12B-3): `_append_enterprise_nodes` depende de
//! `cortex.enterprise.config` (crate cortex-enterprise, tarea P12B-3).
//! Mientras ese crate no exista, si el proyecto TIENE org.yaml se emite
//! warning por stderr y el snapshot queda SIN nodos enterprise (jamás se
//! finge paridad: el gate cubre proyectos sin org.yaml, que es el caso
//! donde Python tampoco agrega nada).

#![forbid(unsafe_code)]

use std::collections::{BTreeSet, HashMap, VecDeque};
use std::path::{Path, PathBuf};

use chrono::Utc;
use serde_json::json;

use crate::cache::WebGraphCache;
use crate::config::WebGraphConfig;
use crate::contracts::{
    WebGraphEdge, WebGraphMode, WebGraphNode, WebGraphNodeDetail, WebGraphSnapshot,
};
use crate::graph_builder::GraphBuilder;
use crate::sources::{EmbedFn, EpisodicSource, SemanticSource};
use cortex_workspace::WorkspaceLayout;

pub struct WebGraphService {
    pub project_root: PathBuf,
    pub config: WebGraphConfig,
    pub semantic_source: SemanticSource,
    pub episodic_source: EpisodicSource,
    pub cache: WebGraphCache,
}

impl WebGraphService {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        project_root: &Path,
        config: Option<WebGraphConfig>,
        vault_path: Option<PathBuf>,
        persist_dir: Option<PathBuf>,
        episodic_entries: Vec<cortex_app::episodic::MemoryEntry>,
        embedder: Option<EmbedFn>,
        workspace_layout: Option<WorkspaceLayout>,
    ) -> WebGraphService {
        let layout_owned;
        let layout: &WorkspaceLayout = if let Some(l) = workspace_layout {
            // Se mueve a un local para poder tomar referencia uniforme.
            layout_owned = l;
            &layout_owned
        } else {
            layout_owned = WorkspaceLayout::discover(project_root);
            &layout_owned
        };
        let config_path = layout.config_path();
        let cfg = config.unwrap_or_else(|| WebGraphConfig::load(Some(project_root), Some(layout)));
        let semantic =
            SemanticSource::new(&config_path, layout, vault_path.clone(), embedder.clone());
        let episodic = EpisodicSource::new(
            &config_path,
            layout,
            persist_dir.clone(),
            episodic_entries,
            embedder,
        );
        let cache = WebGraphCache::new(project_root, Some(layout));
        Self {
            project_root: project_root.to_path_buf(),
            config: cfg,
            semantic_source: semantic,
            episodic_source: episodic,
            cache,
        }
    }

    pub fn build_snapshot_by_mode(
        &self,
        mode: &str,
        use_cache: bool,
        scope: Option<&str>,
    ) -> WebGraphSnapshot {
        let _mode: WebGraphMode = mode.to_string();
        self.build_snapshot(mode, use_cache, scope)
    }

    pub fn build_snapshot(
        &self,
        mode: &str,
        use_cache: bool,
        scope: Option<&str>,
    ) -> WebGraphSnapshot {
        let fingerprint = self.cache.compute_fingerprint(
            &self.semantic_source.vault_path,
            &self.episodic_source.persist_dir,
            self.episodic_source.count(),
            self.episodic_source.cache_token(),
            &self.config.model_dump(),
        );
        if use_cache {
            if let Some(cached) = self.cache.load_snapshot(mode, &fingerprint, scope) {
                return cached;
            }
        }

        let include_embeddings = mode == "hybrid";
        let semantic_records = self.semantic_source.load_records(include_embeddings);
        let episodic_records = self.episodic_source.load_records(include_embeddings);
        let builder = GraphBuilder::new(self.config.clone());
        let generated_at = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Micros, true);
        let mut snapshot = builder.build_snapshot(
            &fingerprint,
            mode,
            &generated_at,
            semantic_records,
            episodic_records,
        );

        // metadata: {"project_id": pid, "scope": "local", **node.metadata}
        let project_id = cortex_workspace::slugify(
            &self
                .project_root
                .file_name()
                .map(|s| s.to_string_lossy().to_string())
                .unwrap_or_default(),
            "project",
        );
        for node in &mut snapshot.nodes {
            let mut merged = serde_json::Map::new();
            merged.insert("project_id".into(), json!(project_id));
            merged.insert("scope".into(), json!("local"));
            for (k, v) in &node.metadata {
                merged.insert(k.clone(), v.clone());
            }
            node.metadata = merged;
        }

        snapshot = append_enterprise_nodes(snapshot, &self.project_root, &project_id);
        snapshot = filter_snapshot_by_scope(snapshot, scope);
        self.cache.store_snapshot(mode, &snapshot, scope);
        snapshot
    }

    pub fn get_node_detail(&self, node_id: &str, mode: &str) -> Option<WebGraphNodeDetail> {
        let snapshot = self.build_snapshot(mode, true, None);
        let nodes_by_id: HashMap<&str, &WebGraphNode> =
            snapshot.nodes.iter().map(|n| (n.id.as_str(), n)).collect();
        let node = nodes_by_id.get(node_id)?;
        let relations: Vec<WebGraphEdge> = snapshot
            .edges
            .iter()
            .filter(|e| e.source == node_id || e.target == node_id)
            .cloned()
            .collect();
        let neighbor_ids: BTreeSet<String> = relations
            .iter()
            .map(|e| {
                if e.source == node_id {
                    e.target.clone()
                } else {
                    e.source.clone()
                }
            })
            .collect();
        let neighbors: Vec<WebGraphNode> = neighbor_ids
            .iter()
            .filter_map(|id| nodes_by_id.get(id.as_str()).map(|n| (*n).clone()))
            .collect();
        Some(WebGraphNodeDetail {
            node: (*node).clone(),
            relations,
            neighbors,
        })
    }

    pub fn resolve_node_path(&self, node_id: &str, mode: &str) -> Option<PathBuf> {
        let detail = self.get_node_detail(node_id, mode)?;
        let rel_path = detail.node.rel_path?;
        if rel_path.is_empty() {
            return None;
        }
        Some(cortex_workspace::layout::resolve_lexical(
            &self.semantic_source.vault_path.join(rel_path),
        ))
    }

    pub fn get_subgraph(
        &self,
        node_id: &str,
        depth: i64,
        mode: &str,
        edge_types: Option<&BTreeSet<String>>,
    ) -> WebGraphSnapshot {
        let snapshot = self.build_snapshot(mode, true, None);
        let mut adjacency: HashMap<String, Vec<(String, String)>> = HashMap::new();
        for edge in &snapshot.edges {
            if let Some(types) = edge_types {
                if !types.contains(&edge.edge_type) {
                    continue;
                }
            }
            adjacency
                .entry(edge.source.clone())
                .or_default()
                .push((edge.target.clone(), edge.edge_type.clone()));
            adjacency
                .entry(edge.target.clone())
                .or_default()
                .push((edge.source.clone(), edge.edge_type.clone()));
        }

        let mut visited: BTreeSet<String> = BTreeSet::new();
        visited.insert(node_id.to_string());
        let mut queue: VecDeque<(String, i64)> = VecDeque::new();
        queue.push_back((node_id.to_string(), 0));
        while let Some((current, current_depth)) = queue.pop_front() {
            if current_depth >= depth {
                continue;
            }
            for (neighbor, _) in adjacency.get(&current).cloned().unwrap_or_default() {
                if visited.contains(&neighbor) {
                    continue;
                }
                visited.insert(neighbor.clone());
                queue.push_back((neighbor, current_depth + 1));
            }
        }

        let nodes: Vec<WebGraphNode> = snapshot
            .nodes
            .iter()
            .filter(|n| visited.contains(&n.id))
            .cloned()
            .collect();
        let edges: Vec<WebGraphEdge> = snapshot
            .edges
            .iter()
            .filter(|e| {
                visited.contains(&e.source)
                    && visited.contains(&e.target)
                    && edge_types.map(|t| t.contains(&e.edge_type)).unwrap_or(true)
            })
            .cloned()
            .collect();
        let mut out = snapshot;
        let node_count = nodes.len() as i64;
        let edge_count = edges.len() as i64;
        out.nodes = nodes;
        out.edges = edges;
        out.stats.node_count = node_count;
        out.stats.edge_count = edge_count;
        out
    }
}

/// Enterprise enrichment — ver GAP en docs del módulo. Con org.yaml presente
/// se declara la dependencia pendiente por stderr (fallo visible, no mudo).
fn append_enterprise_nodes(
    snapshot: WebGraphSnapshot,
    project_root: &Path,
    project_id: &str,
) -> WebGraphSnapshot {
    let layout = WorkspaceLayout::discover(project_root);
    let org_path = layout.org_config_path();
    if !org_path.exists() {
        return snapshot; // caso sin org.yaml: Python también devuelve igual.
    }
    eprintln!(
        "[cortex-webgraph] WARNING: org.yaml presente pero enterprise config \
         aún no portada (P12B-3): nodos enterprise omitidos (project_id={project_id})"
    );
    snapshot
}

fn filter_snapshot_by_scope(
    mut snapshot: WebGraphSnapshot,
    scope: Option<&str>,
) -> WebGraphSnapshot {
    let Some(scope) = scope else {
        return snapshot;
    };
    if scope == "all" {
        return snapshot;
    }
    if scope != "local" && scope != "enterprise" {
        return snapshot;
    }
    let allowed: BTreeSet<String> = snapshot
        .nodes
        .iter()
        .filter(|n| {
            n.metadata
                .get("scope")
                .and_then(|v| v.as_str())
                .unwrap_or("local")
                .trim()
                .to_lowercase()
                == scope
        })
        .map(|n| n.id.clone())
        .collect();
    let nodes: Vec<WebGraphNode> = snapshot
        .nodes
        .iter()
        .filter(|n| allowed.contains(&n.id))
        .cloned()
        .collect();
    let edges: Vec<WebGraphEdge> = snapshot
        .edges
        .iter()
        .filter(|e| allowed.contains(&e.source) && allowed.contains(&e.target))
        .cloned()
        .collect();
    let node_count = nodes.len() as i64;
    let edge_count = edges.len() as i64;
    snapshot.nodes = nodes;
    snapshot.edges = edges;
    snapshot.stats.node_count = node_count;
    snapshot.stats.edge_count = edge_count;
    snapshot
}
